### ISSUE TITLE
Undefined references in libtorch_python.so when linking custom ops in C++

### ISSUE THREAD

**Issue by Masterchef365** [pytorch/pytorch#70882]
### 🐛 Describe the bug

I'm attempting to deploy a model written in Python to a C++ or Rust environment. I used TorchScript to serialize the model, but since the model uses custom ops loading it from C++ or Rust resulted in `Unknown builtin op`. The natural choice here would be to replicate the C++ custom op example [here](https://github.com/pytorch/pytorch/blob/master/test/custom_operator/CMakeLists.txt), which uses cmake to build a custom op and link it with a C++ program. Unfortunately, we are not building our Rust application using CMake. The next route I tried was to use `setup.py` to build a dynamic library, which I would then link with my program. This actually succeeded on Windows, but results in linker errors on Linux. [I found that I could replicate the issue from C++ too, so I made a minimal example available here.](https://github.com/Masterchef365/pytorch_issue_repro) See the repository's README.md for reproduction instructions. Also see the attached log. Thank you!
[log.txt](https://github.com/pytorch/pytorch/files/7818500/log.txt)

### Versions

Collecting environment information...
PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.17

Python version: 3.7.11 (default, Jul 27 2021, 14:32:16)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-91-generic-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: 11.3.109
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080 Ti
Nvidia driver version: 470.86
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.0
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.10.0
[pip3] torchelastic==0.2.0
[pip3] torchtext==0.11.0
[pip3] torchvision==0.11.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.3.1               ha36c431_9    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.3.0           h06a4308_520
[conda] mkl-service               2.4.0            py37h7f8727e_0
[conda] mkl_fft                   1.3.1            py37hd3c417c_0
[conda] mkl_random                1.2.2            py37h51133e4_0
[conda] numpy                     1.21.2           py37h20f2e39_0
[conda] numpy-base                1.21.2           py37h79a1101_0
[conda] pytorch                   1.10.0          py3.7_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchelastic              0.2.0                    pypi_0    pypi
[conda] torchtext                 0.11.0                     py37    pytorch
[conda] torchvision               0.11.0               py37_cu113    pytorch

**Comment by davidberard98** [comment]
@Masterchef365 try linking python in CMakeLists.txt

```cmake
find_package(PythonLibs 3 REQUIRED)
target_link_libraries(test_op "${PYTHON_LIBRARIES}")
```

**Comment by Masterchef365** [comment]
@davidberard98 I'm afraid I couldn't get that to work; I made a branch `python3-libs` in [my repo](https://github.com/Masterchef365/pytorch_issue_repro/tree/python3-libs) to try out that solution. I also tried with specifically the `libpython3.7-dev` package to no avail. Are there other libraries I need to link?

**Comment by davidberard98** [comment]
Is it the same error as before?

**Comment by Masterchef365** [comment]
Yep, here's a log including the container used, the packages installed on it (`libpython3.7-dev`) and with `set( CMAKE_VERBOSE_MAKEFILE on )` so the linker args are shown on line 427. I'm very surprised your suggestion did not work, it seemed to be the right way to go. 
[log.txt](https://github.com/pytorch/pytorch/files/7900944/log.txt)

**Comment by davidberard98** [comment]
Possibly an issue with linker order? Try running the target_link_libraries in different orders. e.g. if `$TORCH_LIBRARIES` depends on python, it should appear before `$PYTHON_LIBRARIES`

**Comment by Masterchef365** [comment]
I swapped the order from `target_link_libraries(test_op "${PYTHON_LIBRARIES}" "${TORCH_LIBRARIES}")` to
`target_link_libraries(test_op "${TORCH_LIBRARIES}" "${PYTHON_LIBRARIES}")` with no effect

**Comment by davidberard98** [comment]
can you also try changing the order relative to `target_link_libraries(test_op add_one)` ?

looking at the most logs, we can see that `-ltorch_python` is linked with `add_one`, and then `-lpython3.7m` appears before `-ladd_one`. I'm thinking that if you can get `-ladd_one` to appear before `-lpython3.7m`, the issue will be solved

**Comment by Masterchef365** [comment]
I've changed the CMakeLists.txt to reflect this, same error:
`target_link_libraries(test_op add_one "${TORCH_LIBRARIES}" "${PYTHON_LIBRARIES}")`.
Resulting linker command:
```
/usr/bin/c++    -rdynamic CMakeFiles/test_op.dir/test_op.cpp.o  -o test_op  -L/cwd/add_one_op/build -Wl,-rpath,/cwd/add_one_op/build:/opt/conda/lib/python3.7/site-packages/torch/lib:/usr/local/cuda/lib64 -ladd_one /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch.so /opt/conda/lib/python3.7/site-packages/torch/lib/libc10.so -lcuda /usr/local/cuda/lib64/libnvrtc.so /usr/local/cuda/lib64/libnvToolsExt.so /usr/local/cuda/lib64/libcudart.so /opt/conda/lib/python3.7/site-packages/torch/lib/libc10_cuda.so -lpython3.7m -Wl,--no-as-needed,"/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cuda.so" -Wl,--as-needed -Wl,--no-as-needed,"/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cuda_cpp.so" -Wl,--as-needed -Wl,--no-as-needed,"/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so" -Wl,--as-needed -lpthread /opt/conda/lib/python3.7/site-packages/torch/lib/libc10_cuda.so /opt/conda/lib/python3.7/site-packages/torch/lib/libc10.so /usr/local/cuda/lib64/libcudart.so /usr/local/cuda/lib64/libnvToolsExt.so /usr/local/cuda/lib64/libcufft.so /usr/local/cuda/lib64/libcurand.so /usr/local/cuda/lib64/libcublas.so -lcudnn -Wl,--no-as-needed,"/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cuda_cu.so" -Wl,--as-needed -Wl,--no-as-needed,"/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch.so" -Wl,--as-needed
```
TL;DR  -ladd_one -lpython3.7m -l\<various torch libs\>

I also tried 
`target_link_libraries(test_op add_one "${TORCH_LIBRARIES}" "${PYTHON_LIBRARIES}")` on the offchance that it works the other way around.

**Comment by davidberard98** [comment]
sorry, not able to repro this on my side right now and don't have many other ideas here

I'd suggest looking into why the linker is still unable to find all those references. You can use `ldd` to find out which libraries/binaries are expecting other libraries, and you can use `nm -DC` to see if the symbols exist in the libraries where they are expected to be found.

**Comment by Masterchef365** [comment]
I've figured out a solution, which is to compile the custom ops using CMake instead of setuptools. An example can be found in the [cmake_custom_op](https://github.com/Masterchef365/pytorch_issue_repro/tree/cmake_custom_op)  branch of my issue repo. Cheers!

**Comment by davidberard98** [comment]
Great, thanks for the update!

**Comment by vbharadwaj-bk** [comment]
Just wanted to mention that I ran into this problem for the same reason (compiled via setuptools from Python rather than CMake) and solved it by linking in libpython from CMake, as recommended.

To achieve the same result using the setuptools from Python, I think you can link in Python using '-L{python_lib_dir}', '-l{python_lib_name}, where

```python
python_lib_dir = sysconfig.get_config_var("LIBDIR")
python_lib_name = sysconfig.get_config_var("LINK_PYTHON_DEPS").replace("lib", "").replace(".a", "")
```
